File: carddemo-common/src/main/java/com/carddemo/common/constant/Messages.java
package com.carddemo.common.constant;

/**
 * Verbatim CardDemo common-message constants (from ``CSMSG01Y``, group
 * ``CCDA-COMMON-MESSAGES``) and abend-diagnostic field widths (from
 * ``CSMSG02Y``, group ``ABEND-DATA``).
 *
 * :purpose: Provide a framework-light, non-instantiable holder for the frozen
 *           display messages and the fixed-width abend-diagnostic area shared
 *           across the CardDemo services.
 * :output:  ``public static final`` message strings preserved byte-identically
 *           to their COBOL source literals, plus abend field-width and
 *           blank-value constants.
 */
public final class Messages {

    /**
     * Non-instantiable constant holder.
     *
     * :purpose: Prevent instantiation of this utility class.
     */
    private Messages() {
    }

    /**
     * Fixed width of the ``CCDA-COMMON-MESSAGES`` fields (``PIC X(50)``).
     */
    public static final int COMMON_MESSAGE_LENGTH = 50;

    // ---------------------------------------------------------------------
    // Common messages (source: CSMSG01Y, group CCDA-COMMON-MESSAGES)
    // ---------------------------------------------------------------------

    /**
     * Sign-off / thank-you message shown on the 3270-equivalent screens.
     *
     * :output: The effective 50-byte value of ``CCDA-MSG-THANK-YOU``. The source
     *          field is declared ``PIC X(50)`` and the 49-character source literal
     *          is right-padded with one trailing space at runtime; this constant
     *          exposes that byte-identical 50-character value.
     */
    public static final String CCDA_MSG_THANK_YOU =
            padToWidth("Thank you for using CardDemo application...      ", COMMON_MESSAGE_LENGTH);

    /**
     * Invalid-key banner shown when an unsupported key is pressed.
     *
     * :output: The effective 50-byte value of ``CCDA-MSG-INVALID-KEY``. The source
     *          field is declared ``PIC X(50)`` and the 49-character source literal
     *          is right-padded with one trailing space at runtime; this constant
     *          exposes that byte-identical 50-character value.
     */
    public static final String CCDA_MSG_INVALID_KEY =
            padToWidth("Invalid key pressed. Please see below...         ", COMMON_MESSAGE_LENGTH);

    // ---------------------------------------------------------------------
    // Abend-diagnostic area (source: CSMSG02Y, group ABEND-DATA)
    //
    // :note: The ABEND-DATA fields are declared VALUE SPACES; only their fixed
    //        field widths carry a contract. The width constants and the
    //        blank-initialized defaults below mirror that layout so downstream
    //        diagnostic code can assemble a fixed-width abend record.
    // ---------------------------------------------------------------------

    /**
     * Width of ``ABEND-CODE`` (``PIC X(4)``).
     */
    public static final int ABEND_CODE_LENGTH = 4;

    /**
     * Width of ``ABEND-CULPRIT`` (``PIC X(8)``).
     */
    public static final int ABEND_CULPRIT_LENGTH = 8;

    /**
     * Width of ``ABEND-REASON`` (``PIC X(50)``).
     */
    public static final int ABEND_REASON_LENGTH = 50;

    /**
     * Width of ``ABEND-MSG`` (``PIC X(72)``).
     */
    public static final int ABEND_MSG_LENGTH = 72;

    /**
     * Blank ``ABEND-CODE`` default (``VALUE SPACES``).
     *
     * :output: A string of ``ABEND_CODE_LENGTH`` spaces.
     */
    public static final String ABEND_CODE = " ".repeat(ABEND_CODE_LENGTH);

    /**
     * Blank ``ABEND-CULPRIT`` default (``VALUE SPACES``).
     *
     * :output: A string of ``ABEND_CULPRIT_LENGTH`` spaces.
     */
    public static final String ABEND_CULPRIT = " ".repeat(ABEND_CULPRIT_LENGTH);

    /**
     * Blank ``ABEND-REASON`` default (``VALUE SPACES``).
     *
     * :output: A string of ``ABEND_REASON_LENGTH`` spaces.
     */
    public static final String ABEND_REASON = " ".repeat(ABEND_REASON_LENGTH);

    /**
     * Blank ``ABEND-MSG`` default (``VALUE SPACES``).
     *
     * :output: A string of ``ABEND_MSG_LENGTH`` spaces.
     */
    public static final String ABEND_MSG = " ".repeat(ABEND_MSG_LENGTH);

    /**
     * Fixed-width serializer mirroring a COBOL alphanumeric ``MOVE`` into a
     * ``PIC X(width)`` field: the value is left-justified and right-padded with
     * spaces to exactly ``width`` characters, or truncated on the right when it is
     * longer than ``width``.
     *
     * :param value: the text to normalize; ``null`` is treated as spaces.
     * :param width: the target fixed field width; must not be negative.
     * :output: a string of exactly ``width`` characters.
     */
    public static String padToWidth(String value, int width) {
        if (value == null) {
            return " ".repeat(width);
        }
        if (value.length() >= width) {
            return value.substring(0, width);
        }
        return value + " ".repeat(width - value.length());
    }
}

File: frontend/src/types/messages.ts
/**
 * :module: messages
 * :purpose: The shared screen-message literals of ``app/cpy/CSMSG01Y.cpy``
 *   (``CCDA-COMMON-MESSAGES``), the copybook every online program includes so the
 *   same two texts appear on every screen. Each literal is published twice: the
 *   fixed-width ``PIC X(50)`` form exactly as the copybook declares it, and the
 *   trimmed form a proportional browser renders.
 * :output: :ts:type:`CCDA_MSG_THANK_YOU_PIC_X50`,
 *   :ts:type:`CCDA_MSG_THANK_YOU`, :ts:type:`CCDA_MSG_INVALID_KEY_PIC_X50`, and
 *   :ts:type:`CCDA_MSG_INVALID_KEY`.
 * :note: The trailing blanks of each ``PIC X(50)`` value are part of the frozen
 *   literal and are preserved verbatim in the ``_PIC_X50`` constants.
 */

/**
 * :purpose: ``CCDA-MSG-THANK-YOU`` verbatim — the text ``COSGN00C`` sends with
 *   ``EXEC CICS SEND TEXT ... ERASE`` when ``DFHPF3`` ends the session (L88-L89).
 */
export const CCDA_MSG_THANK_YOU_PIC_X50 =
  'Thank you for using CardDemo application...       ';

/**
 * :purpose: ``CCDA-MSG-INVALID-KEY`` verbatim — the line-23 text a program moves to
 *   ``WS-MESSAGE`` for an attention identifier it does not handle.
 */
export const CCDA_MSG_INVALID_KEY_PIC_X50 =
  'Invalid key pressed. Please see below...          ';

/** :purpose: Display form of :ts:type:`CCDA_MSG_THANK_YOU_PIC_X50`. */
export const CCDA_MSG_THANK_YOU = CCDA_MSG_THANK_YOU_PIC_X50.trim();

/** :purpose: Display form of :ts:type:`CCDA_MSG_INVALID_KEY_PIC_X50`. */
export const CCDA_MSG_INVALID_KEY = CCDA_MSG_INVALID_KEY_PIC_X50.trim();
